lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -105,49 +153,221 @@ TxOut
 
     -- TxIn
-    utxoTableInputId      TxId        sql=input_tx_id
-    utxoTableInputIndex   Word32      sql=input_index
+    utxoTableInputId         TxId        sql=input_tx_id
+    utxoTableInputIndex      Word32      sql=input_index
 
     -- TxOut
-    utxoTableOutputId     TxId        sql=output_tx_id
-    utxoTableOutputIndex  Word32      sql=output_index
+    utxoTableOutputAddress   W.Address   sql=output_address
+    utxoTableOutputCoin      W.Coin      sql=output_coin
 
     Primary
         utxoTableWalletId
-        utxoTableWalletSlot
+        utxoTableCheckpointSlot
         utxoTableInputId
         utxoTableInputIndex
-        utxoTableOutputId
-        utxoTableOutputIndex
+        utxoTableOutputAddress
+        utxoTableOutputCoin
 
-    Foreign Checkpoint fk_checkpoint_utxo utxoTableWalletId utxoTableWalletSlot
+    Foreign Checkpoint fk_checkpoint_utxo utxoTableWalletId utxoTableCheckpointSlot
     deriving Show Generic
 |]
+
+
+----------------------------------------------------------------------------
+-- Sqlite connection set up
+
+enableForeignKeys :: Sqlite.Connection -> IO ()
+enableForeignKeys conn = stmt >>= void . Sqlite.step
+    where stmt = Sqlite.prepare conn "PRAGMA foreign_keys = ON;"
+
+createSqliteBackend :: Maybe FilePath -> LogFunc -> IO SqlBackend
+createSqliteBackend fp logFunc = do
+  conn <- Sqlite.open (sqliteConnStr fp)
+  enableForeignKeys conn
+  wrapConnection conn logFunc
+
+sqliteConnStr :: Maybe FilePath -> Text
+sqliteConnStr = maybe ":memory:" T.pack
+
+logStderr :: LogFunc
+logStderr _ _ _ str = B8.hPutStrLn stderr (fromLogStr str)
+
+-- | Run a query without error handling. There will be exceptions thrown.
+unsafeRunQuery
+    :: SqlBackend
+    -> SqlPersistM a
+    -> IO a
+unsafeRunQuery conn = runResourceT . runNoLoggingT . flip runSqlConn conn
+
+runQuery
+    :: SqlBackend
+    -> SqlPersistM a
+    -> ExceptT Sqlite.SqliteException IO a
+runQuery conn = ExceptT . try . runResourceT . runNoLoggingT . flip runSqlConn conn
+
+
+----------------------------------------------------------------------------
+-- Database layer methods
+
+-- | Sets up a connection to the SQLite database.
+--
+-- Database migrations are run to create tables if necessary.
+--
+-- If the given file path does not exist, it will be created by the sqlite
+-- library.
+newDBLayer
+    :: Maybe FilePath
+       -- ^ Database file location, or Nothing for in-memory database
+    -> IO (DBLayer IO s t)
+newDBLayer fp = do
+    conn <- createSqliteBackend fp logStderr
+    unsafeRunQuery conn $ runMigration migrateAll
+    return $ DBLayer
+
+        {-----------------------------------------------------------------------
+                                      Wallets
+        -----------------------------------------------------------------------}
+
+        { createWallet = \(PrimaryKey wid) _cp meta ->
+            ExceptT $ unsafeRunQuery conn $
+            selectWallet wid >>= \case
+                Just _ ->
```

Comment:
what would happen if we didn't make this check?

Reply:
if we do not check this then we cannot handle `ErrWalletAlreadyExists` - there are tests that are checking this and my believe is that we do not want to do it

Reply:
There would be an exception from SQLite. Actually I think it will be better to handle this exception than do select + insert.